[net processing] Reduce cs_main scope in MaybeDiscourageAndDisconnect()

jnewbery opened the pull request:
The motivation for this PR is to reduce the scope of cs_main locking in misbehavior logic. It is the first set of commits from a larger branch to move the misbehavior data out of CNodeState and into a new struct that doesn't take cs_main.

There are some very minor behavior changes in this branch, such as:

- Not checking for discouragement/disconnect in `ProcessMessages()` (and instead relying on the following check in `SendMessages()`)
- Checking for discouragement/disconnect as the first action in `SendMessages()` (and not doing ping message sending first)
- Continuing through `SendMessages()` if `MaybeDiscourageAndDisconnect()` doesn't disconnect the peer (rather than dropping out of `SendMessages()`

jnewbery committed:
[net processing] Only call MaybeDiscourageAndDisconnect from SendMessages

`nMisbehavior` is a tally in `CNodeState` that can be incremented from
anywhere. That almost always happens inside a `ProcessMessages()` call
(because we increment the misbehavior score when receiving a bad
messages from a peer), but not always. See, for example, the call to
`MaybePunishNodeForBlock()` inside `BlockChecked()`, which is an
asynchronous callback from the validation interface, executed on the
scheduler thread.

As long as `MaybeDiscourageAndDisconnect()` is called regularly for the
node, then the misbehavior score exceeding the 100 threshold will
eventually result in the peer being punished. It doesn't really matter
where that `MaybeDiscourageAndDisconnect()` happens, but it makes most
sense in `SendMessages()` which is where we do general peer
housekeeping/maintenance.

Therefore, remove the `MaybeDiscourageAndDisconnect()` call in
`ProcessMessages()` and move the `MaybeDiscourageAndDisconnect()` call
in `SendMessages()` to the top of the function. This moves it out of the
cs_main lock scope, so take that lock directly inside
`MaybeDiscourageAndDisconnect()`.

Historic note: `MaybeDiscourageAndDisconnect()` was previously
`SendRejectsAndCheckIfBanned()`, and before that was just sending
rejects.  All of those things required cs_main, which is why
`MaybeDiscourageAndDisconnect()` was called after the ping logic.

jnewbery committed:
[net processing] Change cs_main TRY_LOCK to LOCK in SendMessages()

This was changed to TRY_LOCK in #1117 to fix a potential deadlock
between cs_main and cs_vSend. cs_vSend was split into cs_vSend and
cs_sendProcessing in #9535 (and cs_sendProcessing was changed from a
TRY_LOCK to a LOCK in the same PR).

Since cs_vSend can no longer be taken before cs_main, revert this to a
LOCK().

This commit leaves part of the code with bad indentation. That is fixed
by the next (whitespace change only) commit.

jnewbery committed:
[net processing] Fix bad indentation in SendMessages()

Hint for reviewers: review ignoring whitespace changes.

jnewbery committed:
[net processing] Continue SendMessages processing if not disconnecting peer

If we don't disconnect a peer in MaybeDiscourageAndDisconnect because it
has NOBAN permissions or it's a manual connection, continue SendMessages
processing rather than exiting early.

The previous behaviour was that we'd miss the SendMessages processing on
this iteration of the MessageHandler loop. That's not a problem since
SendMessages() would just be called again on the next iteration, but it
was slightly inefficient and confusing.

MarcoFalke commented on src/net_processing.cpp:
```diff
@@ -3563,32 +3563,44 @@ void ProcessMessage(
-    AssertLockHeld(cs_main);
-    CNodeState &state = *State(pnode.GetId());
+    {
+        LOCK(cs_main);
```
```
net_processing.cpp:3574:9: error: acquiring mutex 'cs_main' that is already held [-Werror,-Wthread-safety-analysis]

        LOCK(cs_main);

        ^

MarcoFalke replied:
You might have to use clang++ to reproduce locally

jnewbery replied:
Ah. I didn't remove the lock annotation from the function declaration. Fixed.

jonatack replied:
Saw this warning as well @ 5722b338d. Now building at 37e726579.

promag commented on src/net_processing.cpp:
```diff
@@ -3881,8 +3883,6 @@ bool PeerLogicValidation::SendMessages(CNode* pto)
     {
         LOCK(cs_main);
 
-        if (MaybeDiscourageAndDisconnect(*pto)) return true;
```
edd1a8a5c76a15174985fe0a1a876eba9e7ee33c

Do you have an idea why this wasn't in the top?

jnewbery replied:
Because it used to be `SendRejectsAndCheckIfBanned()`, and before that it was just sending rejects. Those required cs_main, which is taken after the ping processing at the top.

promag commented on src/net_processing.cpp:
```diff
@@ -3563,32 +3563,44 @@ void ProcessMessage(
+        // Log but don't disconnect
+        LogPrintf("Warning: not punishing whitelisted peer %s!\n", pnode.addr.ToString());
+        return false;
+    } else if (pnode.m_manual_connection) {
```
37e7265796102df35ec83b53fe9114e07d214910

Maybe drop these `else` since you early return on each block.

jnewbery replied:
Agree. That's much clearer.

promag commented on src/net_processing.cpp:
```diff
@@ -3563,32 +3563,44 @@ void ProcessMessage(
-            }
-            connman->DisconnectNode(pnode.addr);
-        }
+    } // cs_main
```
37e7265796102df35ec83b53fe9114e07d214910

nit, I only find this comment useful when the block is big enough.

jnewbery replied:
Agree. That's much clearer.

jnewbery replied:
That's fair, but I don't think it does any harm.

promag commented on src/net_processing.cpp:
```diff
@@ -3563,32 +3563,44 @@ void ProcessMessage(
+    if (pnode.HasPermission(PF_NOBAN)) {
+        // Log but don't disconnect
+        LogPrintf("Warning: not punishing whitelisted peer %s!\n", pnode.addr.ToString());
+        return false;
```
37e7265796102df35ec83b53fe9114e07d214910

No test for this right?

jnewbery replied:
It seems that feature_maxuploadtarget.py tests the `-noban` permission (by verifying that a peer stays connected after exceeding the banscore.

promag reviewed (commented):
Code LGTM 37e7265796102df35ec83b53fe9114e07d214910, some comments though.

jnewbery commented:
Thanks for the review @promag . I've addressed your comments.

jonatack commented on src/net_processing.cpp:
```diff
@@ -3563,32 +3563,48 @@ void ProcessMessage(
+        LogPrintf("Warning: not punishing manually-connected peer %s!\n", pnode.addr.ToString());
+        return false;
+    }
+    
```
The whitespace linter needs appeasing, looks like extra whitespace in lines 3589 and 3595.
```diff
diff --git a/src/net_processing.cpp b/src/net_processing.cpp

@@ -3573,16 +3582,18 @@ bool PeerLogicValidation::MaybeDiscourageAndDisconnect(CNode& pnode)

+

+

^---- failure generated from test/lint/lint-whitespace.sh
```

dongcarl commented on src/net_processing.cpp:
```diff
@@ -3563,32 +3563,48 @@ void ProcessMessage(
+/** Maybe disconnect a peer and discourage future connections from its address.
+ *
+ * @param[in]   pnode           The node to check.
+ * @return                      True if the peer will be disconnected.
```
What about something along the lines of "True if the peer was marked for disconnection in this function"?

jnewbery replied:
Sure. Happy to make that change if I need to retouch this PR.

jonatack replied:
Agree with @dongcarl. Maybe "True if the peer is marked for disconnection in this function."

jnewbery replied:
done

dongcarl reviewed (commented):
Code Review ACK f54af5e58649d527d53990259f77892ad07f7ffa! Happy to see [`2020-06-cs-main-split`](https://github.com/jnewbery/bitcoin/commits/2020-06-cs-main-split) starting to be PR'd :relaxed:

-----

For a minute, I was wondering if it'd be appropriate to move the `MaybeDiscourageAndDisconnect` up to `ThreadMessageHandler`, then realized that would involve changing `NetEventsInterface` so nope.

jonatack commented on src/net_processing.cpp:
```diff
@@ -3563,32 +3563,48 @@ void ProcessMessage(
-    return false;
+
+    // Normal case: Disconnect the peer and discourage all nodes sharing the address
+    LogPrintf("Disconnecting and discouraging peer %s!\n", pnode.addr.ToString());
```
f54af5e nit suggestion while refactoring this function

```diff
@@ -3581,27 +3581,29 @@ bool PeerLogicValidation::MaybeDiscourageAndDisconnect(CNode& pnode)
         state.m_should_discourage = false;
     } // cs_main
 
+    std::string peer{pnode.addr.ToString()};

-        LogPrintf("Warning: not punishing whitelisted peer %s!\n", pnode.addr.ToString());
+        LogPrintf("Warning: not punishing whitelisted peer %s!\n", peer);
 
-        LogPrintf("Warning: not punishing manually-connected peer %s!\n", pnode.addr.ToString());
+        LogPrintf("Warning: not punishing manually-connected peer %s!\n", peer);
 
local address
-        LogPrintf("Warning: disconnecting but not discouraging local peer %s!\n", pnode.addr.ToString());
+        LogPrintf("Warning: disconnecting but not discouraging local peer %s!\n", peer);
 
-    LogPrintf("Disconnecting and discouraging peer %s!\n", pnode.addr.ToString());
+    LogPrintf("Disconnecting and discouraging peer %s!\n", peer);
```

jnewbery replied:
Done (but called the variable `peer_name` since a future commit adds a `Peer` object to this function.

jonatack commented on src/net_processing.cpp:
```diff
@@ -3563,32 +3563,48 @@ void ProcessMessage(
+
+    if (pnode.HasPermission(PF_NOBAN)) {
+        // Peer has the NOBAN permission flag - log but don't disconnect
+        LogPrintf("Warning: not punishing whitelisted peer %s!\n", pnode.addr.ToString());
```
f54af5e58649d527d53990259f77892ad07f7ffa maybe say "NOBAN" or "noban" permission (flag)
```suggestion
        LogPrintf("Warning: not punishing peer with NOBAN permission %s!\n", pnode.addr.ToString());
```

jnewbery replied:
I expect this PR will need to be rebased on #19474, so I've copied the log text from there.

jonatack commented on src/net_processing.cpp:
```diff
@@ -3563,32 +3563,48 @@ void ProcessMessage(
+
+    if (pnode.m_manual_connection) {
+        // Peer is a manual connection - log but don't disconnect
+        LogPrintf("Warning: not punishing manually-connected peer %s!\n", pnode.addr.ToString());
```
f54af5e58649d527d nit while touching this code: s/manually-connected/manually connected/

jnewbery replied:
done

jonatack commented on src/net_processing.cpp:
```diff
@@ -3844,48 +3857,47 @@ class CompareInvMempoolOrder
 
-        // If we get here, the outgoing message serialization version is set and can't change.
-        const CNetMsgMaker msgMaker(pto->GetSendVersion());
+    if (MaybeDiscourageAndDisconnect(*pto)) return true;
```
38c96ef in this commit that removes `MaybeDiscourageAndDisconnect()` from `ProcessMessages()`, future reviewers may appreciate a comment here in the remaining call to make the order dependency explicit.

e.g. that `MaybeDiscourageAndDisconnect()` here depends on all misbehavior calls (currently `Misbehaving`) occurring in ProcessMessages which needs to remain called before SendMessages in `CConnman::ThreadMessageHandler`... if I am not confused.

jnewbery replied:
`nMisbehavior` is a tally in `CNodeState` that can be incremented from anywhere. That almost always happens inside a `ProcessMessages()` call (because we increment the misbehavior score when receiving a bad messages from a peer), but not always. See, for example, the call to `MaybePunishNodeForBlock()` inside `BlockChecked()`, which is an asynchronous callback from the validation interface, executed on the scheduler thread.

As long as `MaybeDiscourageAndDisconnect()` is called regularly for the node, then the misbehavior score exceeding the 100 threshold will eventually result in the peer being punished. It doesn't really matter where that `MaybeDiscourageAndDisconnect()` happens, but it makes most sense in `SendMessages()` which is where we do general peer housekeeping/maintenance.

jonatack replied:
Thanks, @jnewbery, makes sense.

The previous behavior before this PR had me wondering if there were particular optimisation reasons for it, but I don't see any.

jonatack reviewed (commented):
ACK f54af5e58649 with a few comments below.

- Commit message in f54af5e58649: *If we don't disconnect a peer in MaybeDiscourageAndDisconnect because it has NOBAN permissions or it's a manual connection, continue SendMessages processing rather than exiting early* -> Was this a bit of a bug?  EDIT: the former behavior does not seem to have been an anti-DoS optimisation.

- Commit message 38c96ef *Misbehaving()... is only called in ProcessMessages()* -> as it's called from MaybePunishNodeForBlock, MaybePunishNodeForTx, SendBlockTransactions, ProcessHeadersMessage, perhaps write "Misbehaving() calls  originate from callers in ProcessMessages()..."
 
Feel free to ignore the suggestions below.

jnewbery commented:
Thanks for the review @jonatack. I've addressed your inline comments. For your commit log comments:

> Commit message in f54af5e: If we don't disconnect a peer in MaybeDiscourageAndDisconnect because it has NOBAN permissions or it's a manual connection, continue SendMessages processing rather than exiting early -> Was this a bit of a bug?

Not a bug, but inefficient and confusing. I've updated the commit log to include that.

> Commit message 38c96ef Misbehaving()... is only called in ProcessMessages() -> as it's called from MaybePunishNodeForBlock, MaybePunishNodeForTx, SendBlockTransactions, ProcessHeadersMessage, perhaps write "Misbehaving() calls originate from callers in ProcessMessages()..."

I've now realized that this commit log was incorrect (see https://github.com/bitcoin/bitcoin/pull/19472#discussion_r452819554). I've updated the log to be more accurate. Let me know what you think

jnewbery commented:
rebased

MarcoFalke commented on src/net_processing.cpp:
```diff
@@ -3563,32 +3563,50 @@ void ProcessMessage(
+
+    if (pnode.addr.IsLocal()) {
+        // Peer is on a local address. Disconnect this peer, but don't discourage the local address
+        LogPrintf("Warning: disconnecting but not discouraging local peer %s!\n", peer_name);
```
I know this is unrelated, but is there any reason to not log the peer id? All other peer-related messages log the id, so logging the address+port makes it harder to relate this log to all other logs.

practicalswift replied:
Also, isn't the general policy that we should try to avoid logging IP addresses (at least when the user is not opting in to detailed `-debug=net` logging)?

FWIW: IP addresses are considered personal data in some jurisdictions (EU via GDPR?) IIRC.

jnewbery replied:
We've been logging the peer address for node misbehavior since banning was introduced: https://github.com/bitcoin/bitcoin/commit/15f3ad4dbdf0d245b936e68c51a7ecb9f472d2cd#diff-9a82240fe7dfe86564178691cc57f2f1R757 , and these logs have probably just persisted all that time.

I've changed this to log the peer id instead of the address.

jonatack replied:
Wow! -- banscore goes way back, TIL

MarcoFalke replied:
Thanks for fixing!

jonatack commented:
re-ACK a37331f per `git range-diff 505b4ed f54af5e a37331f`

jonatack commented:
re-ACK 655b195 per `git range-diff 505b4ed f54af5e 655b195`, code/commit messages review, a bit of code history, and debug build.

Maybe (only if you have to retouch again):
```diff
 bool PeerLogicValidation::MaybeDiscourageAndDisconnect(CNode& pnode)
 {
-    NodeId peer_id{pnode.GetId()};
+    const NodeId peer_id{pnode.GetId()};
```

MarcoFalke commented on src/net_processing.cpp:
```diff
@@ -3557,32 +3557,49 @@ void ProcessMessage(
+    }
+
+    if (pnode.addr.IsLocal()) {
+        // Peer is on a local address. Disconnect this peer, but don't discourage the local address
```
style-nit in commit 655b195747:

I don't like comments that repeat the source code like `pnode.addr.IsLocal() // Peer is on a local address`. Ideally source code should be self-documenting and in this case simply reading the condition and the logged format string should be sufficient to understand *what* is being done.

However, a comment could make sense to explain *why* something is done. I presume local addresses are not discouraged because all tor-proxy nodes run on the same local address?

(Same for other comments in this function)

I know that this comment has been there before, but since you refactor this function, the comments could be fixed as well.

jnewbery replied:
Done in a follow-up: https://github.com/jnewbery/bitcoin/commit/5a3497f4e8b0d5ce00ecce8779e130b1c4c8ec0c

MarcoFalke commented on src/net_processing.cpp:
```diff
@@ -3557,32 +3557,49 @@ void ProcessMessage(
+    NodeId peer_id{pnode.GetId()};
+    {
+        LOCK(cs_main);
+        CNodeState &state = *State(peer_id);
```
clang-format style-nit in commit 655b1957470c39bcab64917747c9f467444bd809

```suggestion
        CNodeState& state = *State(peer_id);
```

jnewbery replied:
Done in a follow-up: https://github.com/jnewbery/bitcoin/commit/5a3497f4e8b0d5ce00ecce8779e130b1c4c8ec0c

MarcoFalke commented on src/net_processing.cpp:
```diff
@@ -3838,7 +3838,7 @@ class CompareInvMempoolOrder
 bool PeerLogicValidation::SendMessages(CNode* pto)
```
Style-nit in commit body of c581cc16bb5e490c0960ccf440de2d1e5f23c417:

Instead of a GitHub-branded `#1117`, which either needs online-access to GitHub or two local commands to show the referenced commit (`git log --grep='#1117 ' --merges`), what about simply saying `commit c581cc16bb`.

Same for the other one, but since it has two commits, maybe the merge commit can be mentioned `commit 82274c0`

jnewbery replied:
Thanks. Will refer directly to commit hashes in future.

maflcko reviewed (approved):
ACK 655b195747 only some style-nits 🚁

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

ACK 655b195747 only some style-nits 🚁
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUhTIwv/U4oykZ05nHVcYquBiztxhgGyKa9f8uaOypMsTYR6eEWe6L5jcOi7wDYc
0YHHVelm/057+8BOihx6BtYoiStnOzewawMsh2CNXYZPxs9fEJyP5lbTRVLpRN2S
GSwEnXTlrg6N8cZtncmfZRG6XSR9YwwLvS3FXwReTBr9DCE5SPeFz/kUzXin0z2Z
mRG58cYJW4FiYSfRqvd34c3gI8bbJJC9Fz55tc8ZdgXVM9evVeUGQCr1F/cOTbYi
cd0gAkMNbaUuEk3KzO0rK+1iK5oFNo+FCNcU8XG1i1au6tnd+fEBqWOxGJ/TA8VB
m4eI3GHoC4chH4jaOskIYPgSkC5v4ubMV6lImvK3rvud67+5gi/2ZILLxLwjpdCN
KAAKqJTQDGduGgcdt6Tezk1XGmwoJ+CV/sl2dH9INWrEWTsqy9ZQ2sr9uPRPSFva
znujKqFX9MhNkm+tS7XCsRNuJRr3zJN5x4jq9K/tjcCUnfOcCYOJUCmQ2wMpWyBo
2pMOadPJ
=ZRAY
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `cedbe81897bd1c68d1b2c18d961a7b55b239097a710335927096fe5482ada3fd  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e892940108cedbe81897bd1c68d1b2c18d961a7b55b239097a710335927096fe5482ada3fdf0102f2c7d05f5c7e50d0c91f6d81de0452b08fff010963ffa5604fdef26bcf4e47d8f51948108f1045f0983a6f0089e946396178991ec0083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff010a349ab075c50e1f4f54b57ed450a4e5508f1045f0983a7f00883837a0dcd2c814a0083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff010c2d8510c091664c4eaf706d61271d0f908f1045f0983a6f0089c01abffb30b7eda0083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6df010fb145ee7128212185c12f53c3c89498108f1045f0983a6f008149d94caeaba678b0083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6d)

</details>

promag reviewed (commented):
Code review ACK 655b1957470c39bcab64917747c9f467444bd809.

jnewbery commented:
Thanks for the review @MarcoFalke . Rather than invalidate the three ACKs on this branch, I've added a commit to the next branch that addresses your comments: https://github.com/jnewbery/bitcoin/tree/2020-07-tidy-misbehavior.

jnewbery commented:
@jonatack 

> Maybe (only if you have to retouch again):
> 
> bool PeerLogicValidation::MaybeDiscourageAndDisconnect(CNode& pnode)
> ...

Done in follow-up: https://github.com/jnewbery/bitcoin/commits/2020-07-tidy-misbehavior

ariard commented on src/net_processing.cpp:
```diff
@@ -3838,7 +3838,7 @@ class CompareInvMempoolOrder
 {
     const Consensus::Params& consensusParams = Params().GetConsensus();
-    {
+
```
For future work, do we still need `cs_sendProcessing` ? It only guards `m_next_addr_send/m_next_local_addr_send`, both of them only RW in `SendMessages`, inside `ThreadMessageHandler`. 

Is this a relic from the past or do you see another reason ?

jnewbery replied:
I've removed it from `m_next_addr_send` and `m_next_local_addr_send` in https://github.com/jnewbery/bitcoin/commits/2020-06-cs-main-split. It could theoretically also be removed from guarding `SendMessages()` since we only ever call that from the message handler thread.

ariard reviewed (commented):
Code Review ACK 655b195

jnewbery commented:
Thanks for review everyone. The next PR is #19583 (including the suggested fixups by @jonatack and @MarcoFalke)

theuni commented:
I'm worried that we may end up spending a more significant amount of time in SendMessages now, whereas before it was opportunistic. Do you have any idea what the TRY_LOCK hit/miss ratio looked like for a busy node over a few days?

I'd also be curious to know what the cpu usage/profile of a busy node looks like before/after this change.

jnewbery commented:
Thanks for looking at this @theuni!

> Do you have any idea what the TRY_LOCK hit/miss ratio looked like for a busy node over a few days?

I don't know exactly but I'd expect it to be very high (ie we almost always take the lock). What other threads are you expecting to be holding cs_main on a regular basis?

MarcoFalke commented:
I guess this depends on the use case and what other modules are used that might aggressively poll cs_main. (GUI, wallet, indices, RPC, ...)
